Calo towers L1 scouting data format update

tj710 opened the pull request:
#### PR description:

We add new data formats for L1 Scouting, which will this year be collecting the L1 Calo Trigger 'towers', in addition to the contents of the previous years.

#### PR validation:

The tests are documented in /DataFormats/L1Scouting/test/TestL1ScoutingFormat.sh

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

We require backport to CMSSW 16_0_X for data taking purposes from start of 2026 data taking.

cmsbuild commented:
This PR contains too many commits (400 >= 240) and will not be processed.
Please ensure you have selected the correct target branch and consider squashing unnecessary commits.
The processing of this PR will resume once the commit count drops below the limit.

cmsbuild commented:
cms-bot internal usage<!-- bot cache: {"commits":{"158489dbd6bcd3c0d7557af7cd01349f458d82d0":{"files":["DataFormats/L1Scouting/test/TestL1ScoutingFormat.sh","DataFormats/L1Scouting/test/TestReadL1Scouting.cc","DataFormats/L1Scouting/test/TestWriteL1Scouting.cc","DataFormats/L1Scouting/test/create_L1Scouting_test_file_cfg.py","DataFormats/L1Scouting/test/read_L1Scouting_cfg.py"],"squashed":false,"time":1747341064},"94004b77d2f6c41c5b7e546ac987e26bbc3b4b65":{"files":["DataFormats/L1Scouting/interface/OrbitCollection.h"],"squashed":false,"time":1747341067},"c4f70d5932235211bed25c3947c4d625150698e7":{"files":["EventFilter/L1ScoutingRawToDigi/interface/masks.h"],"squashed":false,"time":1747341060},"cb130dec50e6a6b96e4867ec6131d9eaceaa2497":{"files":["DataFormats/L1Scouting/test/TestL1ScoutingFormat.sh","DataFormats/L1ScoutingRawData/interface/SDSNumbering.h"],"squashed":false,"time":1747341073},"d83cd2bbb89772e7b1fdc3a1dfbd751fc7774d96":{"files":["DataFormats/L1Scouting/interface/L1ScoutingCaloTower.h","DataFormats/L1Scouting/src/classes.h","DataFormats/L1Scouting/src/classes_def.xml","DataFormats/L1ScoutingRawData/interface/SDSNumbering.h","EventFilter/L1ScoutingRawToDigi/interface/blocks.h","EventFilter/L1ScoutingRawToDigi/interface/masks.h","EventFilter/L1ScoutingRawToDigi/interface/shifts.h","EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.cc","EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.h"],"squashed":false,"time":1747341050},"e786cd070fe235df112b6737dff0d2971d5e0d2d":{"files":["DataFormats/L1Scouting/interface/L1ScoutingFastJet.h","DataFormats/L1Scouting/src/classes.h","DataFormats/L1Scouting/src/classes_def.xml","L1TriggerScouting/OnlineProcessing/BuildFile.xml","L1TriggerScouting/OnlineProcessing/plugins/BuildFile.xml","L1TriggerScouting/OnlineProcessing/plugins/ScoutingJetProducer.cc","L1TriggerScouting/Utilities/interface/conversion.h","L1TriggerScouting/Utilities/interface/scales.h"],"squashed":false,"time":1747341055}},"emoji":{},"last_seen_sha":"cb130dec50e6a6b96e4867ec6131d9eaceaa2497","signatures":{}} -->

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/44845

ERROR: Build errors found during clang-tidy run.
```
src/L1TriggerScouting/OnlineProcessing/plugins/ScoutingJetProducer.cc:13:10: error: 'FWCore/Utilities/interface/Span.h' file not found [clang-diagnostic-error]
   13 | #include "FWCore/Utilities/interface/Span.h"
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Suppressed 833 warnings (833 in non-user code).
--
gmake: *** [config/SCRAM/GMake/Makefile.coderules:129: code-checks] Error 2
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

missirol commented on DataFormats/L1Scouting/test/read_L1Scouting_cfg.py:
```diff
@@ -29,7 +31,10 @@
   bmtfStubTag = cms.InputTag("l1ScoutingTestProducer", "", "PROD"),
-  expectedBmtfStubValues = cms.vint32(1, 2)
+  expectedBmtfStubValues = cms.vint32(1, 2),
+  caloTowerClassVersion = cms.int32(args.bmtfStubVersion), 
```
Looks wrong (`bmtfStubVersion` instead of `caloTowerVersion`).

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/44855

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1Scouting/test/TestReadL1Scouting.cc  modified in PR(s): #44372
  - File  DataFormats/L1Scouting/test/TestWriteL1Scouting.cc  modified in PR(s): #44372

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/44855/code-format.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/44855/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/44856

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1Scouting/test/TestReadL1Scouting.cc  modified in PR(s): #44372
  - File  DataFormats/L1Scouting/test/TestWriteL1Scouting.cc  modified in PR(s): #44372

cmsbuild commented:
Pull request #48093 was updated.

mmusich commented:
It would be good to update the title to reflect the actual changes proposed instead of the default branch title "Rebased" @tj710

mmusich commented on DataFormats/L1Scouting/test/TestL1ScoutingFormat.sh:
```diff
@@ -33,14 +33,21 @@ cmsRun ${LOCAL_TEST_DIR}/read_L1Scouting_cfg.py --inputFile "$file" || die "Fail
+done
+
+# added Calo tower data format
+oldFiles="testL1Scouting_v3_v3_v3_v3_v3_v3_v3_15_0_1_split_99.root testL1Scouting_v3_v3_v3_v3_v3_v3_v3_15_0_1_split_0.root"
```
how is this supposed to work?
Does it need a new file to be stored in https://github.com/cms-data/DataFormats-L1Scouting ?

tj710 replied:
Hi Marco. Yes, it needs a new file in cms-data, is that repo managed or can I just commit to it? Thanks a lot.

mmusich replied:
Hi Thomas, you need to open a PR to that repo introducing the new file, then the two PRs can be tested together.

tj710 replied:
done 
https://github.com/cms-data/DataFormats-L1Scouting/pull/5

mmusich commented on EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.cc:
```diff
@@ -0,0 +1,120 @@
+  descriptions.addDefault(desc);
+}
+
+DEFINE_FWK_MODULE(ScCaloTowerRawToDigi);
```
missing newline (pedantic)

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/45480

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1Scouting/interface/OrbitCollection.h  modified in PR(s): #48163
  - File  DataFormats/L1Scouting/src/classes.h  modified in PR(s): #48163
  - File  DataFormats/L1Scouting/src/classes_def.xml  modified in PR(s): #48163
  - File  DataFormats/L1Scouting/test/TestReadL1Scouting.cc  modified in PR(s): #44372
  - File  DataFormats/L1Scouting/test/TestWriteL1Scouting.cc  modified in PR(s): #44372

emeschi commented:
@cmsbuild please test with [cms-data/DataFormats-L1Scouting#5](https://github.com/cms-data/DataFormats-L1Scouting/pull/5)

tj710 commented:
Hi @mmusich are you able to issue the required test command? Thank you

BenjaminRS commented:
please test

cmsbuild commented:
-1

**Failed Tests**: UnitTests 
**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa4d55/47233/summary.html
**COMMIT**: d4e42b97e589295ec07d0b8822a11208b5eed0d1
**CMSSW**: CMSSW_15_1_X_2025-07-15-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48093/47233/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Unit Tests

I found 1 errors in the following unit tests:

<pre>---> test TestL1ScoutingFormat had ERRORS
</pre>

## Comparison Summary



**Summary**:
 - You potentially [added 13 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-07-15-1100+fa4d55/69379/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 8 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-07-15-1100+fa4d55/69379/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4076521
 - DQMHistoTests: [Total failures: 70](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-07-15-1100+fa4d55/69379/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4076431
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-07-15-1100+fa4d55/69379/triggerResults): no differences found

quinnanm commented:
please test

gpetruc commented:
@cmsbuild please test with https://github.com/cms-data/DataFormats-L1Scouting/pull/5

gpetruc committed:
code improvements related to L1-Scouting CaloTowers

cmsbuild commented:
+1

**Size**: This PR adds an extra 16KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa4d55/47464/summary.html
**COMMIT**: d4e42b97e589295ec07d0b8822a11208b5eed0d1
**CMSSW**: CMSSW_15_1_X_2025-07-30-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48093/47464/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @valsdav cms-sw/cmsdist#9977

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa4d55/47464/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa4d55/47464/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially [added 16 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-07-30-1100+fa4d55/69613/validateJR/logRootQA-events.log) to the logs
 - [Reco comparison results: 2 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-07-30-1100+fa4d55/69613/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4076162
 - DQMHistoTests: [Total failures: 60](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-07-30-1100+fa4d55/69613/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4076082
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-07-30-1100+fa4d55/69613/triggerResults): no differences found

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_16_0_X. Please open a backport if it should also go in to CMSSW_15_1_X.

quinnanm reviewed (commented):
Just a few minor review comments/questions.

missirol commented on DataFormats/L1Scouting/interface/L1ScoutingCaloTower.h:
```diff
@@ -0,0 +1,38 @@
+    int erBits_;
+    int miscBits_;
+    int hwEta_;
+    int hwPhi_;
```
Every variable in this data format is 32 bits (in most cases), but it looks like the values used to fill these in the unpacker are known to be smaller (for example, `phi` is 8 bits, iiuc). I'm wondering if using smaller types in this data format could lead to a smaller event size. Was that checked ? (or is it irrelevant in this case because compression deals very effectively with all the extra zeros ?)

tj710 replied:
Changed to int16_t

missirol reviewed (commented):
- A question on the data-format types.
 - A possible bug in the CaloTowers' unpacker.
 - A suggestion to change `std::cout` calls to `Log*` calls.

missirol reviewed (commented):
One more comment (see below): it might be necessary to add one more `float` to the `l1ScoutingRun3::FastJet` data format.

tj710 commented on L1TriggerScouting/OnlineProcessing/plugins/ScoutingJetProducer.cc:
```diff
@@ -0,0 +1,114 @@
+#include <utility>
+#include <vector>
+
+using namespace l1ScoutingRun3;
```
addressed

tj710 commented on EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.cc:
```diff
@@ -0,0 +1,120 @@
+    pos += 12;  // header
+
+    if (debug_) {
+      std::cout << " CaloTower #" << sdsId << " Orbit " << orbit << ", BX -> " << bx << ", nCaloTowers -> " << ctCount
```
Addressed

tj710 commented on L1TriggerScouting/OnlineProcessing/plugins/ScoutingJetProducer.cc:
```diff
@@ -0,0 +1,114 @@
+
+  // loop over valid bunch crossings
+  int nConst = 0;
+  double area = 0.;
```
addressed

smorovic commented:
please test

tj710 commented on L1TriggerScouting/OnlineProcessing/plugins/ScoutingJetProducer.cc:
```diff
@@ -0,0 +1,114 @@
+    for (const auto& incJet : incJets) {
+      nConst = incJet.has_constituents() ? incJet.constituents().size() : 0;
+      area = incJet.has_area() ? incJet.area() : -1.0;
+      FastJet fj = FastJet(incJet.Et(), incJet.eta(), incJet.phi(), nConst, area);
```
added mass so we now have
    FastJet(float et, float eta, float phi, int nConst, float area, float mass)

FastJet returns ET rather than PT so went with this. Thanks.

missirol replied:
Just 2 (last) cents: I would put `mass` in fourth place rather than at the end (both in the list of private members of the class, and in the arguments of the constructor). It's much more intuitive to have the 4 four-momentum values close to one another and in that order (in the interest of avoiding future mistakes in the order of constructor arguments, like the case of eta/phi from earlier comments).

tj710 commented on DataFormats/L1Scouting/interface/L1ScoutingCaloTower.h:
```diff
@@ -0,0 +1,38 @@
+    void setErBits(uint16_t erBits) { erBits_ = erBits; }
+    void setMiscBits(uint16_t miscBits) { miscBits_ = miscBits; }
+    void setHwEta(uint16_t hwEta) { hwEta_ = hwEta; }
+    void setHwPhi(uint16_t hwPhi) { hwPhi_ = hwPhi; }
```
Changed all uint16_t to int16_t. Thanks.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/47583

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1Scouting/test/TestReadL1Scouting.cc  modified in PR(s): #44372
  - File  DataFormats/L1Scouting/test/TestWriteL1Scouting.cc  modified in PR(s): #44372

missirol commented on EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.cc:
```diff
@@ -0,0 +1,109 @@
+      erBits = ((ct_raw >> l1ScoutingRun3::calol2::shiftsCaloTowers::erBits) &
+                l1ScoutingRun3::calol2::masksCaloTowers::erBits);
+      miscBits = ((ct_raw >> l1ScoutingRun3::calol2::shiftsCaloTowers::miscBits) &
+                  l1ScoutingRun3::calol2::masksCaloTowers::miscBits);
```
Let's see if I get it (maybe I do not).

 - `erBits` is shifted by 9 and the mask is `0x000e == 1110` , so the selected bits are (10,11,12).
 - `miscBits` is shifted by 12 and the mask is `0x000f == 1111`, so the selected bits are (12,13,14,15).

It looks like bit 12 is used for both. Is that right ? If so, intended ?

missirol reviewed (commented):
Two more comments (hopefully, the last ones).

Thanks for changing the position of `mass` in the jet data format.

tj710 commented on DataFormats/L1Scouting/interface/L1ScoutingCaloTower.h:
```diff
@@ -0,0 +1,38 @@
+#define DataFormats_L1Scouting_L1ScoutingCaloTower_h
+
+#include "DataFormats/L1Scouting/interface/OrbitCollection.h"
+
```
Done. Do you know why the compiler does not complain about this?

tj710 replied:
So should I still add it to this file, or is it not needed? Thanks

RoccoA97 commented on EventFilter/L1ScoutingRawToDigi/interface/shifts.h:
```diff
@@ -104,6 +104,20 @@ namespace l1ScoutingRun3 {
     };
   }  // namespace bmtf
 
+  namespace calol2 {
```
Same as before, maybe this namespace should be `calol1`

tj710 replied:
changed

RoccoA97 commented on EventFilter/L1ScoutingRawToDigi/interface/masks.h:
```diff
@@ -104,6 +104,20 @@ namespace l1ScoutingRun3 {
+      static constexpr uint32_t zeroFlag = 0x0001;
+      static constexpr uint32_t eohrFlag = 0x0002;
+      static constexpr uint32_t hcalFlag = 0x0004;
+      static constexpr uint32_t ecalFlag = 0x0008;
```
Not as relevant as before, but all these masks should be `0x0001`. It is not a big deal as we unpack all of them together in `miscBits` rather than one by one

tj710 commented on EventFilter/L1ScoutingRawToDigi/interface/masks.h:
```diff
@@ -104,6 +104,20 @@ namespace l1ScoutingRun3 {
+  namespace calol2 {
+    struct masksCaloTowers {
+      static constexpr uint32_t ET = 0x01ff;
+      static constexpr uint32_t erBits = 0x000e;
```
fixed

tj710 commented on EventFilter/L1ScoutingRawToDigi/interface/masks.h:
```diff
@@ -104,6 +104,20 @@ namespace l1ScoutingRun3 {
     };
   }  // namespace bmtf
 
+  namespace calol2 {
```
changed

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/47818

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1Scouting/test/TestReadL1Scouting.cc  modified in PR(s): #44372
  - File  DataFormats/L1Scouting/test/TestWriteL1Scouting.cc  modified in PR(s): #44372

tj710 committed:
new data formats for L1 scouting calorimeter trigger towers

Co-Authored-By: Rocco Ardino <rocco.ardino@cern.ch>

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/47838

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1Scouting/test/TestReadL1Scouting.cc  modified in PR(s): #44372
  - File  DataFormats/L1Scouting/test/TestWriteL1Scouting.cc  modified in PR(s): #44372

quinnanm commented on L1TriggerScouting/OnlineProcessing/plugins/ScoutingJetProducer.cc:
```diff
@@ -0,0 +1,114 @@
+#include "L1TriggerScouting/Utilities/interface/conversion.h"
+
+// root libraries
+#include "TLorentzVector.h"
```
This is not used, right? Can it be removed?

quinnanm reviewed (commented):
Hi @tj710 just a few minor comments/questions. When all outstanding comments are addressed can you also already open the 16_0_X backport?

tj710 commented:
> > Hi @tj710 just a few minor comments/questions. When all outstanding comments are addressed can you also already open the 16_0_X backport?
> 
> Just to confirm I backport to 16_0_X not 16_1_X? Thanks

I think yes 16_0. done #50044

quinnanm commented:
thank you!

BenjaminRS commented on EventFilter/L1ScoutingRawToDigi/plugins/ScCALORawToDigi.cc:
```diff
@@ -101,52 +93,43 @@ void ScCaloRawToDigi::unpackOrbitFromDMA(edm::Handle<SDSRawDataCollection>& Scou
-    if (debug_) {
-      std::cout << "CALO Orbit " << orbit << ", BX -> " << bx << std::endl;
-    }
+    LogDebug("L1Scout") << "CALO Orbit " << orbit << ", BX -> " << bx;
```
Think you should have a `<< "\n"` at the end here

BenjaminRS commented on EventFilter/L1ScoutingRawToDigi/plugins/ScCALORawToDigi.cc:
```diff
@@ -185,34 +165,30 @@ void ScCaloRawToDigi::unpackOrbitFromTCP(const unsigned char* buf, size_t len, C
         case CaloObjectType::Jet:
-          if (debug_)
-            std::cout << "Jets BX -> " << bl->bx << std::endl;
+          LogDebug("L1Scout") << "Jets BX -> " << bl->bx;
```
Think you should have a `<< "\n"` at the end here

BenjaminRS commented on EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.cc:
```diff
@@ -0,0 +1,182 @@
+void ScCaloTowerRawToDigi::produce(edm::Event& iEvent, const edm::EventSetup&) {
+  auto const& scoutingRawDataCollection = iEvent.get(rawToken_);
+
+  auto unpackedCaloTowers = std::make_unique<l1ScoutingRun3::CaloTowerOrbitCollection>();
```
Should there be a `new` here?

BenjaminRS commented on EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.cc:
```diff
@@ -0,0 +1,182 @@
+
+    if (bx > NBX) {  // need this check as otherwise the code will crash later accessing orbitBuffer_
+      edm::LogError("ScCaloTowerRawToDigi")
+          << "Corrupt data in sourceId " << sdsId << ", orbit " << orbit << ", invalid BX " << bx;
```
Do we need a `<< "\n"` at the end here?

BenjaminRS commented on EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.cc:
```diff
@@ -0,0 +1,182 @@
+
+    if (debug_) {
+      edm::LogPrint("ScCaloTowerRawToDigi")
+          << " CaloTower #" << sdsId << " Orbit " << orbit << ", BX -> " << bx << ", nCaloTowers -> " << ctCount;
```
Do we need a `<< "\n"` at the end here?

BenjaminRS commented on EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.cc:
```diff
@@ -0,0 +1,182 @@
+      if (debug_) {
+        edm::LogPrint("ScCaloTowerRawToDigi")
+            << "Calo Tower " << i << ", raw: 0x" << std::hex << ct_raw << std::dec << "\n\tET: " << ET
+            << "\n\tER bits: " << erBits << "\n\tMisc bits: " << miscBits << "\n\tEta: " << eta << "\n\tPhi: " << phi;
```
Do we need a `<< "\n"` at the end here?

BenjaminRS reviewed (commented):
Generally everything looks good thanks - just a couple questions on some evaluations in `TestReadL1Scouting.cc`

missirol committed:
L1S: added "tower(Min|Max)HwEt" and "mantissaPrecision" params to ScoutingJetProducer

Added three parameters to the plugin ScoutingJetProducer in order to
(a) apply lower/upper hwEt cuts to the CaloTowers used in jet clustering, and
(b) adjust the precision of the floating-point values used to fill
the data members of instances of the l1ScoutingRun3::FastJet class.

missirol commented on DataFormats/L1Scouting/test/TestReadL1Scouting.cc:
```diff
@@ -361,6 +390,76 @@ namespace edmtest {
+        if (caloTowers[i].hwEta() != expectedCaloTowerValues_[i]) {
+          throwWithMessage("analyzeCaloTowers, hwEta does not match the expected value");
+        }
+        if (caloTowers[i].hwPhi() != expectedCaloTowerValues_[i]) {
```
Since I had to update the PR anyway, in the last push (0d093b48b2b8f42153bc7cd77eae47eb56c54954) I changed this for the new data formats along the lines of what is done for the HLT-Scouting data formats (using different numerical values for different data members, plus using an offset that changes based on event-id and BX-id).

This maybe improves the test a bit. Doing the same for older L1S data formats requires updating files in `cms-data`, and it's beyond the scope of this PR.

cmsbuild commented:
Pull request #48093 was updated. @BenjaminRS, @Dr15Jones, @cmsbuild, @emeschi, @makortel, @quinnanm, @smorovic, @smuzaffar can you please check and sign again.

missirol commented:
The only issue with the the test outputs in https://github.com/cms-sw/cmssw/pull/48093#issuecomment-3883673877 (and https://github.com/cms-sw/cmssw/pull/50044#issuecomment-3883841716) is the following error message:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-fa4d55/51254/dupDict/lostDefs.txt

```
Searching for 'classes_def.xml' in '/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_16_1_X_2026-02-10-2300/src'.

./DataFormats/L1Scouting/src/classes_def.xml defines the following dictionaries that should be defined in another package
  CaloTowers : l1ScoutingRun3::CaloTower
  CaloTowers : vector< l1ScoutingRun3::CaloTower >
  CaloTowers : l1ScoutingRun3::CaloTowerOrbitCollection
  CaloTowers : edm::Wrapper< l1ScoutingRun3::CaloTowerOrbitCollection >
```

In the last push, I updated `duplicateReflexLibrarySearch.py` ([here](https://github.com/cms-sw/cmssw/pull/48093/changes#diff-46a1ee8696bbfa0417aba0da68b26b9e84124da123fc04604e5df5edc69a31f3R39)) following what was done in #29475. Testing this change locally, I don't get the error above anymore. @smuzaffar, FYI (if I should update the script in a different way, please let me know).

Because of this change, the review/signature of @cms-sw/core-l2 is now required.

cmsbuild commented:
Pull request #48093 was updated. @BenjaminRS, @Dr15Jones, @cmsbuild, @emeschi, @makortel, @quinnanm, @smorovic, @smuzaffar can you please check and sign again.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48093/48045

- Found files with invalid states:
  - EventFilter/L1ScoutingRawToDigi/plugins/ScCaloTowerRawToDigi.h:
    - Added: 0d093b48b2b8f42153bc7cd77eae47eb56c54954
    - Deleted: c2919512437e519c620f62d8e6b23da0fb799aa7

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  DataFormats/L1Scouting/test/TestReadL1Scouting.cc  modified in PR(s): #44372
  - File  DataFormats/L1Scouting/test/TestWriteL1Scouting.cc  modified in PR(s): #44372

BenjaminRS commented:
+l1
